tools/src/bin/rust-indexer.rs

```diff
@@ -281,6 +416,24 @@ fn read_existing_contents(
     }
 }
 
+/// Given a reader and a span from that file, extract the text contained by the span.  If the span
+/// covers multiple lines, then whatever newline delimiters the file has will be included.
+fn extract_span_from_source(
+    reader: &mut File,
+    span: &data::SpanData,
+) -> String {
+    reader.seek(std::io::SeekFrom::Start(span.byte_start.into())).unwrap();
+    let len = (span.byte_end - span.byte_start) as usize;
+    let mut buffer: Box<[u8]> = vec![0; len].into_boxed_slice();
+    match reader.read_exact(&mut buffer) {
+        Ok(_) => String::from_utf8(buffer.to_vec()).unwrap(),
```

Comment:
What guarantees that this doesn't fail? This should probably just return a `Result<>` and skip the impl if we fail to read it, or something.

Reply:
Also, `buffer.into_vec()` to avoid copying.

Reply:
I split this fn in two because the Result error types for std::io::Error and std::string::FromUtf8Error didn't line up which made it hard to use `?`.  So the lower-level one retuns io::Result and the outer returns Option<String>.